library/std/src/sys/windows/io.rs

```diff
@@ -78,3 +82,66 @@ impl<'a> IoSliceMut<'a> {
         unsafe { slice::from_raw_parts_mut(self.vec.buf as *mut u8, self.vec.len as usize) }
     }
 }
+
+pub fn is_terminal(h: &impl AsHandle) -> bool {
+    unsafe { handle_is_console(h.as_handle()) }
+}
+
+unsafe fn handle_is_console(handle: BorrowedHandle<'_>) -> bool {
+    let handle = handle.as_raw_handle();
+
+    // A null handle means the process has no console.
+    if handle.is_null() {
+        return false;
+    }
+
+    let mut out = 0;
+    if c::GetConsoleMode(handle, &mut out) != 0 {
+        // False positives aren't possible. If we got a console then we definitely have a console.
+        return true;
+    }
+
+    // At this point, we *could* have a false negative. We can determine that this is a true
+    // negative if we can detect the presence of a console on any of the standard I/O streams. If
+    // another stream has a console, then we know we're in a Windows console and can therefore
+    // trust the negative.
+    for std_handle in [c::STD_INPUT_HANDLE, c::STD_OUTPUT_HANDLE, c::STD_ERROR_HANDLE] {
+        let std_handle = c::GetStdHandle(std_handle);
+        if !std_handle.is_null()
+            && std_handle != handle
+            && c::GetConsoleMode(std_handle, &mut out) != 0
+        {
+            return false;
+        }
+    }
+
+    // Otherwise, we fall back to an msys hack to see if we can detect the presence of a pty.
+    msys_tty_on(handle)
+}
+
+unsafe fn msys_tty_on(handle: c::HANDLE) -> bool {
+    let size = size_of::<c::FILE_NAME_INFO>() + c::MAX_PATH * size_of::<c::WCHAR>();
+    let mut name_info_bytes = vec![0u8; size];
```

Comment:
I've just noticed that this isn't guaranteed to be correctly aligned for a `FILE_NAME_INFO`, since `Vec<u8>` only requests a buffer with an alignment of 1.

In practice the system allocator on windows (and almost every custom allocator that isn't miri, which is unlikely to hit this codepath, and doesn't have a `GetFileInformationByHandleEx` shim even if it did...) will guarantee the needed alignment, so this isn't that big of a deal in practice (we should fix it anyway though).

I added a small struct that simplifies this in https://github.com/rust-lang/rust/pull/101171, so if that lands first (plausible) you can use `Box::new(Align8([0u8; size]))`. If not, doing something equivalent is simple enough.

If it ends up landing without fixing the alignment, I can also pick it up as a quick follow-up once both hit master.

Reply:
That PR landed, so you should be able to use `Align8` (it's in sys/windows/mod.rs, so `crate::sys::Align8`) to ensure this buffer is correctly aligned. It would be nice to get this before landing, but not a big deal if you can't.